FfaFeaturePkg/Library/NotificationServiceLib/NotificationServiceLib.c

```diff
@@ -276,49 +244,52 @@ SetupHandler (
   Service   = NULL;
   ReturnVal = NOTIFICATION_STATUS_NO_MEM;
 
-  /* Extract the UUID from the message */
-  NotificationServiceExtractUuid (Request, Uuid);
-
-  /* If we haven't stored the source and destination IDs, store them */
-  if (!IdsAcquired) {
-    SourceId      = Request->SourceId;
-    DestinationId = Request->DestinationId;
-    IdsAcquired   = TRUE;
-  }
+  /* Extract the UUID from the message x7-x8 (i.e. Arg3-Arg4) */
+  NotificationServiceExtractUuid (Request->Arg3, Request->Arg4, Uuid);
 
   /* Attempt to find the UUID within our list */
   for (Index = 0; Index < NOTIFICATION_MAX_SERVICES; Index++) {
-    if (!CompareMem (Uuid, NotificationServices[Index].ServiceUuid, sizeof (Uuid))) {
-      Service = &NotificationServices[Index];
-      break;
+    /* Check if the service is marked as InUse */
+    if (NotificationServices[Index].InUse) {
+      /* Check if the UUID matches */
+      if (!CompareMem (Uuid, NotificationServices[Index].ServiceUuid, sizeof (Uuid))) {
+        Service = &NotificationServices[Index];
+        break;
+      }
     }
   }
 
   /* The UUID was not found in the list, attempt to find an empty location to add it */
   if (Service == NULL) {
     for (Index = 0; Index < NOTIFICATION_MAX_SERVICES; Index++) {
       if (!NotificationServices[Index].InUse) {
+        /* Temporarily point to the empty location */
         Service = &NotificationServices[Index];
-        CopyMem (Service->ServiceUuid, Uuid, sizeof (Uuid));
-        Service->InUse = TRUE;
         break;
       }
     }
   }
 
-  /* Check for a valid UUID and validate the input parameters */
+  /* Check for a valid UUID */
   if (Service != NULL) {
-    ReturnVal = UpdateServiceBits (FALSE, Request, Service);
+    ReturnVal = UpdateServiceInfo (FALSE, Request, Service);
+    /* Check if the update was successful and this was a new addition */
+    if ((ReturnVal == NOTIFICATION_STATUS_SUCCESS) && (!Service->InUse)) {
+      /* Update the UUID and set the location to InUse */
+      CopyMem (Service->ServiceUuid, Uuid, sizeof (Uuid));
+      Service->InUse = TRUE;
+    }
   } else {
-    DEBUG ((DEBUG_ERROR, "Service Setup Failed - Error Code: %x\n", ReturnVal));
+    DEBUG ((DEBUG_ERROR, "Service Register Failed - Error Code: %d\n", ReturnVal));
   }
 
-  Response->Arg0 = ReturnVal;
+  /* Update the error code */
+  Response->Arg6 = ReturnVal;

```

Comment:
This should be truncated to 8 bits. Otherwise, you are you will be setting some of the reserved bits.

![image](https://github.com/user-attachments/assets/26add2a1-156a-4910-a949-7badda87e081)